### PR TITLE
CLI gateway server address default and mismatch check

### LINKED ISSUE #637: The CLI does not set the gateway server address of new gateways
#### Summary

Just like we do with the network/application/join server address end devices, we need to register the gateway server address for a newly created, and make sure we don't get connection stats

#### Steps to Reproduce

1. Create a gateway
2. Get the gateway server address of the gateway

#### What do you see now?

It's empty

#### What do you want to see instead?

It should be the one that was configured in the CLI, just as we do with end devices.

#### How do you propose to implement this?

- Rename [`compareServerAddresses`](https://github.com/TheThingsNetwork/lorawan-stack/blob/master/cmd/ttn-lw-cli/commands/end_devices.go#L745) so that it contains EndDevice and create a similar function for gateways.
- This function should be called before [calling the gateway server](https://github.com/TheThingsNetwork/lorawan-stack/blob/master/cmd/ttn-lw-cli/commands/gateways.go#L341) for getting the connection stats
- On gateway create, set the gateway server address [just as we do for end devices](https://github.com/TheThingsNetwork/lorawan-stack/blob/master/cmd/ttn-lw-cli/commands/end_devices.go#L312)

#### Can you do this yourself and submit a Pull Request?

Sure, but I believe @furtiman was looking for a nice LoRaWAN Stack (or Go in general) issue.

### PR DESCRIPTION
**Pull request by furtiman**
#### Summary

Closes #637 

#### Changes

- Added default (configured for CLI) gateway server address assignation to a gateway created with `gateways create` without specified `--gateway-server-adress`
- Added gateway server address mismatch check in `connection-stats` command

#### Notes for Reviewers

As requested by @htdvisser . The same procedures happen to end-devices' `ns`, `js`, `as`, so there is no reason gateways should not have it for `gs`.

Creating gateway without `--gateway-server-address`, `localhost` set as default:
```
go run ./cmd/ttn-lw-cli gateways create gtw5 \
  --user-id admin \
  --frequency-plan-id EU_863_870 \
  --gateway-eui 00800000A01009BA \
  --enforce-duty-cycle 
grpc_service=ttn.lorawan.v3.GatewayRegistry namespace=grpc
{
  "ids": {
    "gateway_id": "gtw5",
    "eui": "00800000A01009BA"
  },
  "created_at": "2019-05-15T21:59:07.302Z",
  "updated_at": "2019-05-15T21:59:07.302Z",
  "version_ids": {

  },
  "gateway_server_address": "localhost",
  "frequency_plan_id": "EU_863_870",
  "antennas": [
    {
      "location": {
        "source": "SOURCE_REGISTRY"
      }
    }
  ],
  "enforce_duty_cycle": true
```

Creating a gateway with specifying the `--gateway-server-address`:
```
go run ./cmd/ttn-lw-cli gateways create gtw6 \
  --user-id admin \
  --frequency-plan-id EU_863_870 \
  --gateway-eui 00800000A01009B0 \
  --enforce-duty-cycle \
--gateway-server-address 1234567
 DEBUG Using access token (valid until 12:39PM)
grpc_service=ttn.lorawan.v3.GatewayRegistry namespace=grpc
{
  "ids": {
    "gateway_id": "gtw6",
    "eui": "00800000A01009B0"
  },
  "created_at": "2019-05-16T09:40:22.381Z",
  "updated_at": "2019-05-16T09:40:22.381Z",
  "version_ids": {

  },
  "gateway_server_address": "1234567",
  "frequency_plan_id": "EU_863_870",
  "antennas": [
    {
      "location": {
        "source": "SOURCE_REGISTRY"
      }
    }
  ],
  "enforce_duty_cycle": true
}
```

Mismatch check, gtw6's g-s-a '1234567' differs from CLI's `localhost`, which leads to warning:

```
go run ./cmd/ttn-lw-cli gateways connection-stats "gtw6"
grpc_service=ttn.lorawan.v3.GatewayRegistry namespace=grpc
  WARN Registered Gateway Server address does not match CLI configuration configured=localhost registered=1234567
error:cmd/ttn-lw-cli/commands:gs_address_mismatch (gateway server address mismatch)
exit status 255
```

### CODE REVIEW

**Comment by furtiman** [comment]
I don't really understand why this happens with Travis:

```
$ make git.diff
Previous operations have created changes that were not recorded in the repository. Please make those changes on your local machine before pushing them to the repository:
diff --git a/config/messages.json b/config/messages.json
index 1edb29a..fd69361 100644
--- a/config/messages.json
+++ b/config/messages.json
@@ -1187,6 +1187,15 @@
       "file": "end_devices.go"
     }
   },
+  "error:cmd/ttn-lw-cli/commands:gs_address_mismatch": {
+    "translations": {
+      "en": "gateway server address mismatch"
+    },
+    "description": {
+      "package": "cmd/ttn-lw-cli/commands",
+      "file": "gateways.go"
+    }
+  },
   "error:cmd/ttn-lw-cli/commands:inconsistent_end_device_eui": {
     "translations": {
       "en": "given end device EUIs do not match registered EUIs"
make: *** [git.diff] Error 1
The command "make git.diff" exited with 2.
```

Can someone give me a hint?

**Comment by KrishnaIyer** [comment]
@furtiman: run `mage go:messages`

**Comment by furtiman** [comment]
@KrishnaIyer 
```
./mage go:messages
Language: en	Translations:   687	Missing:     0
```

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 1]
```diff
@@ -740,9 +740,9 @@ func init() {
 	Root.AddCommand(endDevicesCommand)
 }
 
-var errAddressMismatch = errors.DefineAborted("address_mismatch", "server address mismatch")
+var errAddressMismatchEndDevice = errors.DefineAborted("address_mismatch", "server address mismatch")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Let’s use either the same error, or make sure that the error name reflects the scope

**Comment by furtiman** [comment]
2 identical Travis builds, one stuck?

**Comment by johanstokking** [comment]
I restarted both

**Comment by furtiman** [comment]
From what I found online this is Travis' issue, sometimes it starts two identical PR builds, one of which gets infinitely queued.

**Comment by johanstokking** [comment]
In that case an admin needs to merge

**Comment by furtiman** [comment]
Either way I am not authorized to merge yet 👶

File: cmd/ttn-lw-cli/commands/gateways.go [hunk 2]
```diff
@@ -377,6 +408,7 @@ func init() {
 	gatewaysCreateCommand.Flags().AddFlagSet(setGatewayFlags)
 	gatewaysCreateCommand.Flags().AddFlagSet(setGatewayAntennaFlags)
 	gatewaysCreateCommand.Flags().AddFlagSet(attributesFlags())
+	gatewaysCreateCommand.Flags().Bool("defaults", true, "configure gateway with default gateway server address")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/gateways.go]: @johanstokking don't you think that while we're at it, we should make the defaults similar to end devices? The default options are, same as with the end devices, options that most users would want on almost all devices, such as `enforce-duty-cycle`.

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 3]
```diff
@@ -740,9 +740,9 @@ func init() {
 	Root.AddCommand(endDevicesCommand)
 }
 
-var errAddressMismatch = errors.DefineAborted("address_mismatch", "server address mismatch")
+var errAddressMismatchEndDevice = errors.DefineAborted("end_device_address_mismatch", "network/application/join server address mismatch")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/end_devices.go]: The error ID should be `end_device_server_address_mismatch` (it's not about the end device address).

File: cmd/ttn-lw-cli/commands/gateways.go [hunk 4]
```diff
@@ -394,3 +426,17 @@ func init() {
 	gatewaysCommand.AddCommand(gatewaysContactInfoCommand)
 	Root.AddCommand(gatewaysCommand)
 }
+
+var errAddressMismatchGateway = errors.DefineAborted("gs_address_mismatch", "gateway server address mismatch")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/gateways.go]: Error ID should be `gateway_server_address_mismatch` for consistency with end devices.

Comment (furtiman) [reply cmd/ttn-lw-cli/commands/gateways.go]: corrected

File: cmd/ttn-lw-cli/commands/end_devices.go (hunk 3 above)
Comment (furtiman) [reply cmd/ttn-lw-cli/commands/end_devices.go]: corrected

File: cmd/ttn-lw-cli/commands/gateways.go (hunk 2 above)
Comment (furtiman) [reply cmd/ttn-lw-cli/commands/gateways.go]: @htdvisser I believe that for end devices the defaults are also set only for [ns/as](https://github.com/TheThingsNetwork/lorawan-stack/blob/92a8b210b6516228785e35239ed7e1e2dc455e45/cmd/ttn-lw-cli/commands/end_devices.go#L310) and [js](https://github.com/TheThingsNetwork/lorawan-stack/blob/92a8b210b6516228785e35239ed7e1e2dc455e45/cmd/ttn-lw-cli/commands/end_devices.go#L359), correct me if I'm wrong

**Comment by furtiman** [comment]
Also Travis issue is gone, you can assign me back if you want to

**Comment by furtiman** [comment]
😞 
Screenshot 2019-05-17 at 16 01 18

File: cmd/ttn-lw-cli/commands/gateways.go (hunk 2 above)
Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/gateways.go]: @htdvisser agreed. I would propose;

- `auto_update` on
- `enforce_duty_cycle` on
- `status_public` on
- `location_public` on

**Comment by furtiman** [comment]
@johanstokking any defaults to add for end-devices then, if we are at that?

**Comment by johanstokking** [comment]
@furtiman no those are good for now

**Comment by furtiman** [comment]
Added defaults mentioned by @johanstokking 

```
go run ./cmd/ttn-lw-cli gateways create gtw7 \
  --user-id admin \
  --frequency-plan-id EU_863_870 \
  --gateway-eui 10800000A01009B0

grpc_service=ttn.lorawan.v3.GatewayRegistry namespace=grpc
{
  "ids": {
    "gateway_id": "gtw7",
    "eui": "10800000A01009B0"
  },
  "created_at": "2019-05-20T21:24:57.805Z",
  "updated_at": "2019-05-20T21:24:57.805Z",
  "version_ids": {

  },
  "gateway_server_address": "localhost",
  "auto_update": true,
  "frequency_plan_id": "EU_863_870",
  "antennas": [
    {
      "location": {
        "source": "SOURCE_REGISTRY"
      }
    }
  ],
  "status_public": true,
  "location_public": true,
  "enforce_duty_cycle": true
}
```

File: cmd/ttn-lw-cli/commands/gateways.go [hunk 5]
```diff
@@ -202,9 +205,27 @@ var (
 					return err
 				}
 			}
+
+			setDefaults, _ := cmd.Flags().GetBool("defaults")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/gateways.go]: This is fine for now, but for the future: if you only need `setDefaults` once, we prefer writing it as

```go
if setDefaults, _ := cmd.Flags().GetBool("defaults"); setDefaults { 
    ...
```

File: cmd/ttn-lw-cli/commands/gateways.go (hunk 2 above)
Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/gateways.go]: The comment for this flag needs to be updated.

Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/gateways.go]: Also Gateway Server is written in title case

File: cmd/ttn-lw-cli/commands/gateways.go (hunk 5 above)
Comment (furtiman) [reply cmd/ttn-lw-cli/commands/gateways.go]: For remembering the correct practices, I corrected it as well.

**Comment by furtiman** [comment]
Updated according to the feedback/

**Comment by johanstokking** [comment]
@furtiman you have merge rights now

**Pull request merged**